AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/TgzPackageExtract.kt

```diff
@@ -0,0 +1,330 @@
+
+    val BUFFER = 512
+    val TOO_BIG_SIZE: Long = 0x6400000 // max size of unzipped data, 100MB
+    val TOO_MANY_FILES = 1024 // max number of files
+
+    var count = 0
+    var total: Long = 0
+    var data = ByteArray(BUFFER)
+
+    /**
+     * Determine whether a file is a gzip.
+     *
+     * @param file the file to check.
+     * @return whether the file is a gzip.
+     * @throws IOException if the file could not be read.
+     */
+    @Throws(IOException::class)
+    fun isGzip(file: File?): Boolean {
+        val signature = ByteArray(GZIP_SIGNATURE.size)
+        FileInputStream(file).use { stream ->
+            if (stream.read(signature) != signature.size) {
+                return false
+            }
+        }
+        return GZIP_SIGNATURE.contentEquals(signature)
+    }
+
+    /**
+     * Untar and ungzip a tar.gz file to a AnkiDroid/addons directory.
+     *
+     * @param tarballFile the .tgz file to extract
+     * @param addonsDir   the AnkiDroid addons directory
+     * @return the temp directory.
+     * @throws FileNotFoundException if .tgz file or ungzipped file i.e. .tar file not found
+     * @throws IOException
+     */
+    @Throws(Exception::class)
+    fun extractTarGzipToAddonFolder(tarballFile: File, addonsDir: File) {
+
+        require(isGzip(tarballFile)) { context.getString(R.string.not_valid_js_addon, tarballFile.absolutePath) }
+
+        if (!addonsDir.exists()) {
+            context.getString(R.string.could_not_create_dir, addonsDir.absolutePath)
+        }
+
+        // Make sure we have 2x the tar file size in free space (1x for tar file, 1x for unarchived tar file contents
+        requiredMinSpace = tarballFile.length() * 2
+        availableSpace = Utils.determineBytesAvailable(addonsDir.canonicalPath)
+        if (requiredMinSpace > availableSpace) {
+            Timber.e("Not enough space to extract file, need %d, available %d", requiredMinSpace, availableSpace)
+            throw IOException(context.getString(R.string.import_log_insufficient_space_error, Formatter.formatFileSize(context, requiredMinSpace), Formatter.formatFileSize(context, availableSpace)))
+        }
+
+        // If space available then unGZip it
+        val tarTempFile = unGzip(tarballFile, addonsDir)
+        tarTempFile.deleteOnExit()
+
+        // Make sure we have sufficient free space
+        val unTarSize = calculateUnTarSize(tarTempFile)
+        if (unTarSize > availableSpace) {
+            Timber.e("Not enough space to untar, need %d, available %d", unTarSize, availableSpace)
+            throw IOException(context.getString(R.string.import_log_insufficient_space_error, Formatter.formatFileSize(context, unTarSize), Formatter.formatFileSize(context, availableSpace)))
+        }
+
+        try {
+            // If space available then unTar it
+            unTar(tarTempFile, addonsDir)
+        } catch (e: IOException) {
+            Timber.w("Failed to unTar file")
+        }
```

Comment:
we know that we are done working with the tarTempFile now, and it may have been huge relative to available space, let's be kind to the user and delete the file immediately. The deleteOnExit is just to make sure that we delete even if we somehow go in background and are killed by system, we can still be proactive here

```suggestion
        try {
            // If space available then unTar it
            unTar(tarTempFile, addonsDir)
        } catch (e: IOException) {
            Timber.w("Failed to unTar file")
        } finally {
            tarTempFile.delete();
        }
```